utilities/transactions/transaction_test.h

```diff
@@ -0,0 +1,145 @@
+
+#include "port/port.h"
+
+namespace rocksdb {
+
+class TransactionTest : public ::testing::TestWithParam<
+                            std::tuple<bool, bool, TxnDBWritePolicy>> {
+ public:
+  TransactionDB* db;
+  FaultInjectionTestEnv* env;
+  std::string dbname;
+  Options options;
+
+  TransactionDBOptions txn_db_options;
+
+  TransactionTest() {
+    options.create_if_missing = true;
+    options.max_write_buffer_number = 2;
+    options.write_buffer_size = 4 * 1024;
+    options.level0_file_num_compaction_trigger = 2;
+    options.merge_operator = MergeOperators::CreateFromStringId("stringappend");
+    env = new FaultInjectionTestEnv(Env::Default());
+    options.env = env;
+    options.concurrent_prepare = std::get<1>(GetParam());
+    dbname = test::TmpDir() + "/transaction_testdb";
+
+    DestroyDB(dbname, options);
+    txn_db_options.transaction_lock_timeout = 0;
+    txn_db_options.default_lock_timeout = 0;
+    txn_db_options.write_policy = std::get<2>(GetParam());
+    Status s;
+    if (std::get<0>(GetParam()) == false) {
+      s = TransactionDB::Open(options, txn_db_options, dbname, &db);
+    } else {
+      s = OpenWithStackableDB();
+    }
+    assert(s.ok());
+  }
+
+  ~TransactionTest() {
+    delete db;
+    DestroyDB(dbname, options);
+    delete env;
+  }
+
+  Status ReOpenNoDelete() {
+    delete db;
+    db = nullptr;
+    env->AssertNoOpenFile();
+    env->DropUnsyncedFileData();
+    env->ResetState();
+    Status s;
+    if (std::get<0>(GetParam()) == false) {
+      s = TransactionDB::Open(options, txn_db_options, dbname, &db);
+    } else {
+      s = OpenWithStackableDB();
+    }
+    return s;
+  }
+
+  Status ReOpen() {
+    delete db;
+    DestroyDB(dbname, options);
+    Status s;
+    if (std::get<0>(GetParam()) == false) {
+      s = TransactionDB::Open(options, txn_db_options, dbname, &db);
+    } else {
+      s = OpenWithStackableDB();
+    }
+    return s;
+  }
+
+  Status OpenWithStackableDB() {
+    std::vector<size_t> compaction_enabled_cf_indices;
+    std::vector<ColumnFamilyDescriptor> column_families{ColumnFamilyDescriptor(
+        kDefaultColumnFamilyName, ColumnFamilyOptions(options))};
+
+    TransactionDB::PrepareWrap(&options, &column_families,
+                               &compaction_enabled_cf_indices);
+    std::vector<ColumnFamilyHandle*> handles;
+    DB* root_db;
+    Options options_copy(options);
+    Status s =
+        DB::Open(options_copy, dbname, column_families, &handles, &root_db);
+    if (s.ok()) {
+      assert(handles.size() == 1);
+      s = TransactionDB::WrapStackableDB(
+          new StackableDB(root_db), txn_db_options,
+          compaction_enabled_cf_indices, handles, &db);
+      delete handles[0];
+    }
+    return s;
+  }
+};
+
+class MySQLStyleTransactionTest : public TransactionTest {};
+
+// TODO(myabandeh): Instantiate the tests with other write policies
+INSTANTIATE_TEST_CASE_P(DBAsBaseDB, TransactionTest,
```

Comment:
These INSTANTIATE_TEST_CASE_P should be in .cc?

Reply:
Sure.